### ISSUE TITLE
Add automatic resizing of partitions

### ISSUE THREAD

**Issue by shasheene** [rescuezilla/rescuezilla#18]
A recurring [feature request](https://sourceforge.net/p/redobackup/discussion/1169663/thread/77385e35/) is automatic resizing of partitions depending on the size of the destination drive.

Clonezilla supports this, so technically the feature will be achieved with #4

**Comment by shasheene** [comment]
It sounds like Clonezilla's "-r" argument can [only](https://clonezilla.org/clonezilla-live/doc/03_Disk_to_disk_clone/advanced/05-advanced-param.php) be used to grow the target partition:

> If the target disk size is larger than the source disk, you can try to use option "-k1" which will create the partition table proportionally in the target disk and turn on option "-r" to resize the file file system in the partition automatically. This is useful to make use all of the target disk size.

Also from the Clonezilla FAQ:

>  [Why Clonezilla can _NOT_ restore an image which is saved from a large drive to a smaller drive? Any workaround?](https://drbl.org/faq/fine-print.php?path=./2_System/25_restore_larger_disk_to_smaller_one.faq#25_restore_larger_disk_to_smaller_one.faq)
> --
> It's not easy to implement such a feature, since Clonezilla now is a partition "imaging" tool, by "imaging", it means clonezilla actually does not know the files themselves, clonezilla just knows where the used blocks are. Because of this reason, the target partition size must be equal or larger than the original one so that clonezilla can restore the used blocks on that partition. If the target partition size is smaller, then it will go wrong.Unless Clonezilla has file-based function in the future. Maybe...If you really still want to make this, check this FAQ, the doc here, this forum or the program fsarchiver.

and:

[Is that possible I can restore to smaller HDD when the image was taken from bigger one? Or to clone a larger disk to a smaller one?](https://drbl.org/faq/fine-print.php?path=./2_System/119_larger_to_smaller_disk_restore_clone.faq#119_larger_to_smaller_disk_restore_clone.faq)
--
This is always a challenge to Clonezilla because it's block-based tool, not file-based one.Two possible solutions:Method 1: If you are sure the used data blocks on the source disk are within the disk boundary of destination one, yes, it's possible to make that with the options "-k1" and "-icds" enabled in the expert mode.Method 2: Use Clonezilla live >= 2.6.7-28 or 20200703-focal. For older version it does not always work.First use GParted live to resize the source disk, and make all partitions to be equal to or smaller than the size of destination disk.Then boot Clonezilla live.For disk to disk, follow this.For image to disk, follow this.The only difference is, just remember to choose expert mode, and enable "-icds" (while do not choose any option about "-k*").

Given Rescuezilla's recent addition in [supporting](https://github.com/rescuezilla/rescuezilla/issues/20) Image Explorer (beta) support, there might be a (complex) pathway to implement this. However this doesn't sound like a great solution.

**Comment by candrapersada** [comment]
I mean, individual image backup and restore should be "swap" and "extended" select or deselected for individual backups?

**Comment by shasheene** [comment]
Deleted comment that was a duplicate of [this comment](https://github.com/rescuezilla/rescuezilla/issues/145#issuecomment-753432566)

**Comment by shasheene** [comment]
Growing partitions is easy to implement. All Rescuezilla would need to do is restore the partition table and partclone backup, then use the correct programs to grow the partition and filesystem.

**One of the key features of FOG Project is the ability to restore to disks _smaller_ than original.** FOG Project mostly relies on partclone (like the industry-standard Clonezilla format that Rescuezilla uses) but uses different metadata so [FOG Project images have only recently become supported by Rescuezilla](https://github.com/rescuezilla/rescuezilla/issues/194). I have noticed that FOG Project images often have a modified sfdisk partition table backup, which presumably provides flexibility in how the partition table is restored. According to the FOG Project wiki describing its [Single Disk - Resizable image format](https://wiki.fogproject.org/wiki/index.php/Managing_FOG#Creating_Images_Objects), it appears FOG Project's default settings is to _shrink_ the partitions _before_ using partclone to make a backup:

> This is the default choice used by FOG as it works in most cases and allows for deployment to smaller size disks as well. It takes a copy of every partition on the disk, and resizes partitions that has excessive free space to a smaller size where possible. Each resizable partition will go through a "Resizing filesystem" process for each partition that is to be resized. This process can take some time depending on how severe disk fragmentation is. The partitions that are shrank will be shrunk down to only 2GB of free space on their partition. This allows an image taken from a 6TB drive with only 20GB of used space to be deployable to a drive with a total capacity of 25GB roughly. When the partitions are laid onto the destination drive, all resized partitions are intelligently expanded to utilize the entire drive.

Given Rescuezilla's [primary goal](https://github.com/rescuezilla/rescuezilla/wiki/Rescuezilla-Project-Roadmap) is to be fully compatible with the industry-standard Clonezilla format, **it does not make sense for Rescuezilla generate backups uses the exact approach taken by FOG Project**, because custom logic will be needed to expand the images which Clonezilla is unlikely to implement which would break Rescuezilla's full Clonezilla interoperability. Also because [Rescuezilla supports all the partclone-based imaging formats (Clonezilla, Redo Backup and Recovery, FOG Project, Redo Rescue and Foxclone)](https://github.com/rescuezilla/rescuezilla/issues/194), **almost all input images will not be pre-shruken**. So a different approach to FOG Project is required.

So for deploying non FOG Project images, it may be possible to stream the output of partclone into the relevant filesystem resizing application, but this relies on piping the output between applications. Given streams do not provide random access this may not be easy. Side note: I need to further investigate how FOG Project is able to pre-shrink images.

Right now the approach of mounting the image using partclone-utils may be the cleanest solution, and would exercise a code path that gets tested widely with Rescuezilla Image Explorer. It will of course first [require a lot of work to make partclone-utils more robust](https://sourceforge.net/p/partclone-utils/tickets/3/).

**Comment by shasheene** [comment]
For completeness, Clonezilla's implementation in **growing** the filesystem is found in `sbin/ocs-resize-part`. Here are the filesystem supported:
* reiserfs (via `/sbin/resize_reiserfs` of reiserfsprogs package)
* vfat|fat16|fat32 (`/usr/sbin/fatresize` from fatresize package)
* ext2|ext3|ext4 (via `/sbin/resize2fs` of e2fsprogs package)
* ntfs (via `/sbin/ntfsresize` of ntfs-3g package)
* xfs (via `/usr/sbin/xfs_growfs` of xfsprogs package)
* jfs (via `mount -o remount,resize`
* btrfs (via `btrfs filesystem resize max`)
* nilfs2 (via `/usr/sbin/nilfs-resize` of nilfs-tools package)

As best I can tell, Clonezilla doesn't first modify the partition table other than relying on a partition table proportional restore.

For Rescuezilla v2.2 it doesn't make sense to implement this `sbin/ocs-resize-part` behavior yet, as achieving the ability to automatically grow _and_ shrink the filesystems across all supported image formats is much more useful.

**Comment by shasheene** [comment]
The FOG Project implementation of the shrinking feature described in the previous comment is a very inefficient and roundabout way that makes both the backup and restore operations  _really_ slow. I have some ideas on how to implement this key feature much more efficiently:

1. Fix remaining [problems in the partclone-utils](https://sourceforge.net/p/partclone-utils/tickets/3/) project (which will also improve Image Explorer (beta)’s low-compatibility)
    * This will unlock the ability of any partclone image to be expressed as a regular block device
2. Either update each filesystem resize program like fatresize, ntfsresize (see previous comment) to handle arbitrary destination devices, rather than only supporting in-place resizing
    * This will be a lot of work, but it should be the most efficient as the program can understands it won't need to do temporary copies like the in-place case.
3. Or alternatively, implement a kind of copy-on-write layer using Network Block Device to abstract resizing without needing to add a new feature to the resize program for every supported filesystem.
    * This will be much less work, but the program could be slower due to the move operations made by the resize programs causing unnecessary temporary copies, unless the NBD layer intelligently handles this.

Either approach requires fixing partclone-utils. Despite all the work it will require, the payoff seems worthwhile: automatic resizing when restoring or cloning to disks of different capacity to the original will be a powerful feature.

**Comment by mzky** [comment]
I have tried to restore the system to a small hard disk using the following methods

Backup system:
```
tar Pzcfp backup.tgz --exclude=/proc --exclude=/mnt --exclude=/lost+found --exclude=/sys --exclude=/media --exclude=/run --exclude=/boot --exclude=/dev --exclude=/var/spool/postfix /
```

Restore the system：
By starting the U disk system (LiveCD), manually partition the disk, and decompress the system backup files (About 3-5 minutes)
```
tar zxf backup.tgz
```
Repair GRUB :
```
blkid 
```
Populate GRUB with the obtained UUID:
```
vim /boot/grub/grub.conf
root="obtained UUID"

vim /etc/fstab
uuid="obtained UUID"
 
Under the Centos7 system, it can be used normally.

**Comment by shasheene** [comment]
Exactly. The `backup.tgz` that you describe is an archive of all _files_ on your system, and the approach you describe involves partitioning the disk, copying the _files_ in and then modifying GRUB bootloader.

**This is the approach I will implement for Rescuezilla as described by option 4**. It uses partclone-utils to mount the backup image's existing partclone files before doing any copying, so that backups created by any support image format automatically become able to be restored to disk smaller than original.

(This feature is something that no other open-source disk imaging and cloning tool has yet implemented. Not even Clonezilla. FOG Project's roundabout implementation of this feature using "Single Disk - Resizable image format" is too slow to be useful, so doesn't count.)

**Comment by alphanis** [comment]
It will be great if Rescuezilla support this feature. I think Foxclone does, according to its manual page 23. [https://foxclone.org/downloads/foxcloneV51.pdf]

**Comment by shasheene** [comment]
@alphanis Foxclone appears to only support this with the [same manual workaround as the current version of Rescuezilla](https://github.com/rescuezilla/rescuezilla/wiki/HOWTO:-Restoring-to-a-smaller-disk.-Eg,-1000GB-HDD-to-500GB-SSD). From that project's manual:

![Screenshot_2024-02-01_20-20-17](https://github.com/rescuezilla/rescuezilla/assets/1414379/9870b729-b64e-4ee4-8289-b8f639cf6914)

**Comment by snipeTR** [comment]
Hello,

I’ve been using Rescuezilla with great appreciation for a long time. However, this issue (**#18 – “Add automatic resizing of partitions**”), which has been open since 2019 and was specifically noted as being worked on in 2020, still has not been resolved. In 2025, even in the age of artificial intelligence, the fact that we still don’t have a method to automatically resize partitions intelligently and transfer the system completely to the target disk seems like a serious shortcoming.

Of course, I understand that this feature is technically not easy to implement. But if the main purpose of Rescuezilla is to save the user from manual steps and simplify the process, then it’s time to take this step. What users expect is the ability to move their system not only to larger disks, but also to smaller or differently sized ones smoothly, with all necessary adjustments done automatically. Moreover, with the undeniable progress of internet speeds and cloud systems, we need large disks far less than before.

I would also like to ask: how about collaborating with Microsoft to ensure the proper migration of Windows operating systems? Having official guidance on which files need modification, which steps are critical, and how to guarantee a cloned system boots correctly on the same machine would be very valuable. Modern operating systems involve highly complex security checks and intricate configurations. That’s why such collaboration could be an important step toward strengthening Rescuezilla’s Windows compatibility.

In short, I kindly but strongly request that this feature be prioritized in future versions of Rescuezilla. Without it, it’s difficult to speak of a truly “fully automatic” recovery/cloning experience.

**Thank you, and I remain grateful for all your work.**
